cr.swift

```diff
@@ -0,0 +1,41 @@
+//
+//  Netguru iOS code review task
+//
+
+class paymentViewController: UIViewController {
+    var delegate: PaymentViewControllerDelegate
+    let customView = PaymentView()
+    let payment: Payment?
+
+    func callbacks() {
+        view.backgroundColor = UIColor(displayP3Red: 1, green: 1, blue: 1, alpha: 1)
+        customView.didTapButton = {
+            if let payment = self.payment {
+                self.delegate.didFinishFlow(amount: payment.amount
+                                            currency: payment.currency)
+            }
+        }
+    }
+
+    func viewDidLoad() {
+        navigationController?.setNavigationBarHidden(false, animated: false)
+        customView.centerXAnchor.constraint(equalTo: view.centerXAnchor)
+        customView.centerYAnchor.constraint(equalTo: view.centerYAnchor)
+        view.addSubview(customView)
+        fetchPayment()
+        setupCallbacks()
+    }
+
+    internal let ViewModel: PaymentViewModel
+
+    func fetchPayment() {
+        customView.statusText = "Fetching data"
+        ApiClient.sharedInstance().fetchPayment { payment in
+            self.CustomView.isEuro = payment.currency == "EUR" ? true : false
+            if payment!.amount != 0 {
+                self.CustomView.label.text = "\(payment!.amount)"
+                return
+            }
+        }
+    }
```

Comment:
- Try to don't overuse **Singelton**, if necessary to use singleton then hide it after abstraction and inject during initialisation to allow testability and to decouple code
- singleton pattern has static let shared instance not functions returning an instance
- when using async operation use **[weak self]** to avoid retain cycles
- additionally, all this kind of operation should be moved to **ViewModel** and **ApiClient** should be injected there, then the ViewModel should return only formated for the view data that controller will use to update views
- don't force unwrap payment, it's unsafe when you are not sure that there will be value

Reply:
- inside async operation closures please update UI in the main thread